viamonvif/device/device.go

```diff
@@ -0,0 +1,326 @@
+	var resp GetProfilesResponseEnvelope
+	err = xml.NewDecoder(bytes.NewReader(b)).Decode(&resp)
+	if err != nil {
+		return zero, fmt.Errorf("failed to decode media profiles response: %w", err)
+	}
+
+	if len(resp.Body.GetProfilesResponse.Profiles) == 0 {
+		dev.logger.Warn("No media profiles found in the response")
+		return zero, errors.New("no media profiles found")
+	}
+
+	dev.logger.Debugf("Found %d media profiles", len(resp.Body.GetProfilesResponse.Profiles))
+	for i, profile := range resp.Body.GetProfilesResponse.Profiles {
+		dev.logger.Debugf("Profile %d: Token=%s, Name=%s", i, profile.Token, profile.Name)
+	}
+
+	dev.logger.Debugf("GetProfiles decoded: %#v", resp.Body.GetProfilesResponse)
+	return resp.Body.GetProfilesResponse, nil
+}
+
+type getStreamURIResponse struct {
+	XMLName xml.Name `xml:"Envelope"`
+	Body    struct {
+		GetStreamURIResponse struct {
+			MediaURI onvif.MediaUri `xml:"MediaUri"`
+		} `xml:"GetStreamUriResponse"`
+	} `xml:"Body"`
+}
+
+// Credentials contain an onvif device username and password.
+type Credentials struct {
+	User string `json:"user"`
+	Pass string `json:"pass"`
+}
+
+// GetStreamURI returns a device's stream URI for a given profile.
+func (dev *Device) GetStreamURI(profile onvif.Profile, creds Credentials) (*url.URL, error) {
+	dev.logger.Debugf("GetStreamUri token: %s, profile: %#v", profile.Token, profile)
+	body, err := dev.callMedia(GetStreamURI{
+		StreamSetup: onvif.StreamSetup{
+			Stream:    onvif.StreamType(streamTypeRTPUnicast),
+			Transport: onvif.Transport{Protocol: streamSetupProtocol},
+		},
+		ProfileToken: profile.Token,
+	})
+	if err != nil {
+		return nil, err
+	}
+	dev.logger.Debugf("GetStreamUri response: %v", string(body))
+
+	var streamURI getStreamURIResponse
+	if err := xml.NewDecoder(bytes.NewReader(body)).Decode(&streamURI); err != nil {
+		return nil, fmt.Errorf("failed to get RTSP URL for profile %s: %w", profile.Token, err)
+	}
+
+	dev.logger.Debugf("GetStreamUriResponse decoded %v: ", profile.Token, streamURI)
+
+	uriStr := string(streamURI.Body.GetStreamURIResponse.MediaURI.Uri)
+	if uriStr == "" {
+		return nil, fmt.Errorf("got empty uri for profile %s", profile.Token)
+	}
+
+	uri, err := url.Parse(uriStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI %s: %w", uriStr, err)
+	}
+
+	if creds.User != "" || creds.Pass != "" {
+		uri.User = url.UserPassword(creds.User, creds.Pass)
+	}
+	return uri, nil
+}
+
+// GetEndpoint returns specific ONVIF service endpoint address.
+func (dev *Device) GetEndpoint(name string) string {
+	return dev.endpoints[name]
+}
+
+func (dev Device) callMedia(method interface{}) ([]byte, error) {
+	return dev.callMethodDo(dev.endpoints["media"], method)
+}
+
+func (dev Device) callDevice(method interface{}) ([]byte, error) {
+	return dev.callMethodDo(dev.endpoints["device"], method)
+}
+
+func (dev Device) callMethodDo(endpoint string, method interface{}) ([]byte, error) {
```

Comment:
can we give this method a more descriptive name

Reply:
how is that?